presto-main/src/main/java/com/facebook/presto/operator/aggregation/OptimizedTypedSet.java

```diff
@@ -0,0 +1,198 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.block.DictionaryBlock;
+import com.facebook.presto.common.block.DictionaryId;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.operator.project.SelectedPositions;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static com.facebook.presto.operator.project.SelectedPositions.positionsList;
+import static com.facebook.presto.type.TypeUtils.hashPosition;
+import static com.facebook.presto.type.TypeUtils.positionEqualsPosition;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static it.unimi.dsi.fastutil.HashCommon.arraySize;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class OptimizedTypedSet
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(TypedSet.class).instanceSize();
+    private static final float FILL_RATIO = 0.75f;
+    private static final int EMPTY_SLOT = -1;
+
+    private final Type elementType;
+    private final String functionName;
+
+    // size is the number of elements added to the TypedSet (including null).
+    private int size;
+    private int hashCapacity;
+    private int hashMask;
+
+    private Block[] blocks;
+    private int maxBlockCount;
+    private List<SelectedPositions> positionsForBlocks = new ArrayList<>();
+    private long[] blockPositionByHash;  // Each 64-bit long is 32-bit index for blocks + 32-bit position within block
+    private int currentBlockIndex = -1;
+
+    public OptimizedTypedSet(Type elementType, int maxBlockCount, int maxPositionCount, String functionName)
+    {
+        checkArgument(maxPositionCount >= 0, "maxPositionCount must not be negative");
+        this.elementType = requireNonNull(elementType, "elementType must not be null");
+        this.functionName = functionName;
+        this.hashCapacity = arraySize(maxPositionCount, FILL_RATIO);
+        this.hashMask = hashCapacity - 1;
+
+        this.maxBlockCount = maxBlockCount;
+        blocks = new Block[maxBlockCount];
+        blockPositionByHash = new long[hashCapacity];
+        Arrays.fill(blockPositionByHash, EMPTY_SLOT);
+    }
+
+    /**
+     * Add the block into the hash table blockPositionByHash using "union" semantics
+     */
+    public void addUnion(Block block)
+    {
+        if (positionsForBlocks.size() == maxBlockCount) {
+            throw new IllegalArgumentException("Cannot add block to OptimizedTypedSet: maxBlockCount is reached");
+        }
+
+        int positionCount = block.getPositionCount();
+
+        currentBlockIndex++;
+        blocks[currentBlockIndex] = block;
```

Comment:
we can do ```blocks[++currentBlockIndex] = block;```

Reply:
I believe using ++ on the same line is not preferred by Presto. But it does save one line.